### ISSUE TITLE
stdout newline conversion of binary data

### ISSUE THREAD

**Issue by daanzu** [kkm000/openfst#20]
On Windows, outputting binary FST data to stdout from `fstcompile` (and I assume others) appears to perform newline conversion, inserting 0x0d before every 0x0a and rendering the data invalid. I've yet to look deeply into why. Do you know of a easy solution?

**Comment by daanzu** [comment]
FYI, I fixed this issue, at least for my use cases. The changes are small, but spread out. @kkm000 Do you want a PR, or is it not important enough?

**Comment by jtrmal** [comment]
I think Kyril is busy but will definitely want the PR :)
y.

**Comment by brucelin** [comment]
> FYI, I fixed this issue, at least for my use cases. The changes are small, but spread out. @kkm000 Do you want a PR, or is it not important enough?

I may meet the same problem when I use kaldi under the cygwin enviroment, can u show your solution. Thank you very much. Below is my case
```
utils/make_lexicon_fst.pl --pron-probs $tmpdir/lexiconp.txt $sil_prob $silphone | \
    fstcompile --isymbols=$dir/phones.txt --osymbols=$dir/words.txt \
    --keep_isymbols=false --keep_osymbols=false | \
     fstarcsort --sort_type=olabel > $dir/L.fst || exit 1;
```
 and the error information 
```
ERROR: GenericRegister::GetEntry: No such file or directory
ERROR: MutableFst::Read: Unknown FST type "vector" (arc type = "standard"): standard input
```

**Comment by daanzu** [comment]
Oops, I've been lax in sending this in. I'll try to post it today.

**Comment by daanzu** [comment]
@brucelin Should be fixed in #23. Also, some binaries with this fix are available if you download and unpack (with zip) the wheels for https://github.com/daanzu/kaldi-active-grammar.

**Comment by brucelin** [comment]
@daanzu Thank you very much. I will try it.

**Comment by kkm000** [comment]
Thanks for discovering the issue and the patch, and I'm sorry--was really busy last couple months.

I remember there was a way to deal with the text/binary streams using [`std::freopen`](https://en.cppreference.com/w/cpp/io/c/freopen), do you think it would be a better approach if it works?

> If `filename` is a null pointer, then the function attempts to reopen the file that is already associated with stream (it is implementation defined which mode changes are allowed in this case) 

So it should be like `std::freopen(nullptr, "wb", stdout);`. I recall it did the trick.

**Comment by daanzu** [comment]
Neat, I don't recall knowing about the `std::freopen` approach. It's been awhile since I wrote the patch, so I can't remember exactly why I chose this approach. I can say that it appears to have been working perfectly for me for the past 7 months, however.

**Comment by kkm000** [comment]
It's supposed to be more portable, although it's not required by the standard and left to the implementation as to which mode can or cannot be changed. But IIRC, it worked with MSVC. I'll check and let you know.

**Comment by kkm000** [comment]
Blocked-By: #30

**Comment by kkm000** [comment]
Repro:
```
$ printf '0 1 10 10\n1\n' | fstcompile | xxd -u -s64
```
Linux output (ok):

00000040: 0000 0000 807F 0100 0000 0000 0000 **0A**00
00000050: 0000 **0A**00 0000 0000 0000 0100 0000 0000
00000060: 0000 0000 0000 0000 0000

Windows output (bug):

00000040: 0000 0000 807F 0100 0000 0000 0000 **0D0A**
00000050: 0000 00**0D 0A**00 0000 0000 0000 0100 0000
00000060: 0000 0000 0000 0000 0000 0000

**Comment by kkm000** [comment]
Repro 2. On a Windows machine, save the binary generated by fstcompile:
```
$  printf '0 1 10 10\n1\n'
0 1 10 10
1
$ printf '0 1 10 10\n1\n' | fstcompile | gzip | base64 -w0                                                             
H4sIAAAAAAAAA7v2d1MdGwMDQ1lqckl+EQeQVVySmJeSWJTCxAABzAyNUVkMCMDEgA4a6hmhLF4uGAECjGgqAUmO7J5sAAAA
```
Switch to Linux terminal, and do the reverse, unencode, ungzip, fstprint:
```
$ echo 'H4sIAAAAAAAAA7v2d1MdGwMDQ1lqckl+EQeQVVySmJeSWJTCxAABzAyNUVkMCMDEgA4a6hmhLF4uGAECjGgqAUmO7J5sAAAA' | base64 -d | gunzip | ./fstprint
0       65536   2573    658688
1
```
The result is nonsense.

**Comment by kkm000** [comment]
Fixed in #32
